### DISCUSSION TITLE
Switch source code license to GPL

### DISCUSSION THREAD

**Discussion by pastk** [organicmaps/organicmaps#6003] [Ideas]
Currently Organic Maps sources are under "permissive" Apache 2.0 license.

Let's discuss pros and cons of switching to GPL or some other "copyleft" license.

The main idea is to guarantee that the project will stay open source.
(e.g. OM's parent Maps.Me was open source, but now is closed/proprietary).

**Comment by hb0nd** [comment]
I firmly agree with this 👍

i don't see a downside for OM, copyleft will prevent fragmentation, and ensure the project will last well into the future (hopefully indefinitely)

**Comment by pastk** [comment]
Another advantage is that under GPL it will be legally impossible to make a derived product based on OM and keep it closed (e.g commercial). The one who did it will be legally obliged to open source the changes to the code, so that OM and a wider community will be able to use enhancements as well.

**Comment by pastk** [comment]
Possible disadvantages/restrictions:
- it'll be impossible (or very hard at least) to change the license again;
- it'll be impossible to dual-license, so less monetization possibilities.

**Reply by muralito** [reply]
I fully support GPL license, but it seems some of your ideas are somehow contradictory.

At first lance your goal is "guarantee that the project will stay open source" but you bring as a con that it could be impossible to dual -license it. This is exactly what happened with Maps.me, they have the rights of their code and stop releasing it as open source, and continued developing it under a proprietary license.

It's near impossible if the rights are fragmented between multiple authors, but it's still possible to dual license in GPL and to any other of license if OM have the rights. Every author has to transfer the IP rights of their contributions to the OM organization. Do that and then, OM could convert in the "new Maps.me", stop releasing OrganicMaps as opensource, and release only under a proprietary license.

What are the risks for OrganicMaps? What do you don't want that third parties can do with your code? What do you want to allow?
Could a third party (e.g .Maps.me) take portions of OrganicMaps and improve their product and continue relasing it as closed source? Do you want that?

**Reply by pastk** [reply]
I think there is a concern indeed, that non-free projects could use OM's improvements without contributing back.

I've listed impossibility to dual-license as a **possible** disadvantage, because I don't really know what is the stance of OM's co-founders who "hold the project's keys" on this GPL proposition.

It would be great to know what do you think, guys? @rtsisyk @biodranik @vng 

Do you keep in mind a possibility to finance the project via dual-licensing and don't want to close this door?
Or any other concerns with GPL?

**Reply by FedericoCeratto** [reply]
In copyright law original authors retain the right to use any license they want in future releases.
Regarding past releases both Apache and GPL grant non-revocable rights to users - aka authors cannot "unrelease" a past release.
GPL does not (cannot) prohibit dual-licensing.

If there are many contributors it can be difficult to reach an agreement for example to provide a special license to a company (but this is not a problem specific to GPL)

**Reply by pastk** [reply]
well yeap that's what I meant exactly - under the GPL to provide a special license OM will have to get consent of all the contributors which is very hard, while under Apache its a non-issue

**Reply by FedericoCeratto** [reply]
If your intention is to provide exceptions in exchange for payment this is not currently viable using Apache. Instead, it is possible for future contributors to sign an agreement to allow the project to make this specific form of dual licensing (GPL or paid) viable.
Unlike other CLAs, such agreement would not necessarily require contributors to give away authorship or allow arbitrary license changes, but only allow the specific use-case you have in mind.

**Comment by RicoElectrico** [comment]
Disadvantage: it would be hard to spin off components as libraries. Currently every high-level functionality (map rendering, search, routing) is a custom job. But making a loosely coupled library out of some of them could recoup some of the development effort and hopefully attract contributors among people who would use the library.
For now, I think the most viable element to librarize would be routing. It's quite good and calculates routes quickly. While voice instructions are so-so, there isn't much competition in the space. Hardcoded C++-based profiles make it very fast on mobile. Other use scripting languages which are too slow on phones.

**Reply by pastk** [reply]
For libraries there is LGPL, right?
Or you mean it won't be possible to switch parts of the code from GPL to LGPL?

**Reply by mpawelski** [reply]
Yeah, I'm not open-source license expert (in fact, I'm a total noob in this 😅), but I I fear that switching to more copy-left license could prevent any "librarification" initiatives in the future.
 
I'm not sure that there is any need or wish to do that. I just think it would be cool if OM map could be easily embed in a mobile app. Even cooler, imagine if OM could be embedded on a website, that OM could be another layer on official https://osm.org/ website 🤩. Sounds, cool right? I'm not sure if this is even technically possible, I know OM code is mostly in C++ and OpenGL for rendering map, and this stuff can run in browser with WASM (WebAssembly). And licenses like GPL might not be "web friendly", I even hear that LGPL doesn't even make sense in web word (what is "dynamic linking" of C++ libraries equivalence in JS/web world?). Just some food for thought from my side.

**Reply by pastk** [reply]
Why GPL is not "web friendly"? Actually AFAIK its kind of looses its copyleft power when used for a web service (because there is no copying or distribution of software). And that is why there is AGPL to preserve copyleft for SaaS things :)

**Reply by FedericoCeratto** [reply]
Components can be spun off as library at any time under the same (GPL or AGPL) license and they would be still useful (most of the mapping ecosystem is GPL anyways). I'd be happy to upload them into Debian and Ubuntu as library packages.

**Comment by pastk** [comment]
Also I have no idea if the change might possibly affect non-code assets like artwork, translations, etc...
E.g. we use icons under following licenses: https://github.com/organicmaps/organicmaps/blob/98f814037537b836e65fa77f6ee6debe26b42359/data/copyright.html#L300-L342

**Reply by FedericoCeratto** [reply]
As long as the copyright belongs to Google, Fonticons and Aleksey Popov those licenses cannot be changed by the Organic Maps authors. Those licenses would stay the same (and they are compatible with GPL).

**Comment by pastk** [comment]
It'd be great if someone who knows the matter better will comment about feasibility and possible tricky unobvious consequences (if any)...

**Comment by spaetz** [comment]
I am not a lawyer, but a researcher who has been dealing with licenses for 2 decades now. On the question of which licensed code can be incorporated/changed into which licenses, I really like the license compatability chart in Figure 2.2. at https://wiki.geant.org/pages/viewpage.action?pageId=725614690#SoftwarelicenceselectionandmanagementinG%C3%89ANT-2.3CompatibilityofLicencesFrequentlyUsedinG%C3%89ANT which shows that Apache code can easily be incorporated into a GPL project (but not the other way around).
Also, yes, libraries can e.g. be released under LGPL which is "weaker" on the copyleft aspects.

As for other assets, there is no need to put everything under a single license, e.g. GPL does not really make sense for fonts and icons, so it is fine to have those under different licenses, especially if they are more liberally licensed.

**Comment by FedericoCeratto** [comment]
If the project was going to move to GPL or LGPL I'd be happy to contribute and maintain license tracking files according to the https://reuse.software/ standard to provide clarity to users and developers.

**Comment by pastk** [comment]
Actually our Android builds (except F-Droid ones) include a proprietary Google Fused Location Services client library (https://developers.google.com/location-context/fused-location-provider/) which is used to smartly combine location data from gps, wifi, cell bluetooth sources..

Only in Android 12 it became possible to use this service via a standard open-source Android Location API, see https://source.android.com/docs/automotive/location/coarse-location

So if OM switches to GPL then it won't be possible to ship this proprietary lib.

**Reply by pastk** [reply]
Maybe there is an interim option to have the C++ core GPLed while keeping Android code under Apache 2.0?

Shipping C++ native code and using it via JNI is considered static or dynamic linking?

**Reply by pastk** [reply]
Also I want to share my experience living without Google Fused Location Services (GFLS).

I used to have an ungoogled phone with microG. It was noticeably slower to obtain a location fix compared to my friend's stock android phone. So I thought that using GFLS makes a significant difference.

A few months ago I upgraded my phone (all phones mentioned are Xiaomi btw) and also degooged it (its Divested OS now), but no microG this time. I.e. supposedly there is no network location in my new phone.
Still it gets location fixes much-much faster than my previous one with microG. Its a tad slower still compared to my friend's stock phone, but the difference is so minor now that I'm not bothered.

So maybe my old microG setup was broken, or there were some A-GPS issues, or maybe phone's hardware has a more profound effect on location fixing speed. Maybe if I install microG on my new phone then it'll be completely on par with stock using GFSL..

My current understanding is that GFSL advantages might be overrated.

**Reply by Atemu** [reply]
> proprietary Google Fused Location Services client library (https://developers.google.com/location-context/fused-location-provider/)

There is a Free re-implementation by the µG project. Linking https://github.com/organicmaps/organicmaps/pull/9575 here which intends to use that for f-droid flavours as that'd be acceptable on F-droid, albeit likely with an anti-feature.

> Maybe there is an interim option to have the C++ core GPLed while keeping Android code under Apache 2.0?
> 
> Shipping C++ native code and using it via JNI is considered static or dynamic linking?

That wouldn't work and no, it wouldn't make a difference as far as the GPL is concerned. If it's part of the software, it must be distributed under the terms of the GPL and if the dependency's license isn't compatible, that won't work.

IANAL though.

> So maybe my old microG setup was broken, or there were some A-GPS issues, or maybe phone's hardware has a more profound effect on location fixing speed

If the phone is old enough, it may not have had support for some of the "newer" constellations such as the EU's Galileo. I believe Android uses all the constellations that the hardware can work at once with to deduce its location. In apps like GPSTest you can see which constallations' satellites are used for your current location.

I don't know how the details of this work but if your old device only had access to Navstar's GPS, it may have needed to wait a little longer for enough satellites to be in view or something. My FP4 gets fixes much quicker and more reliably than my old cheeseburger which could only do GPS and GLONASS IIRC.

**Reply by muralito** [reply]
> Actually our Android builds (except F-Droid ones) include a proprietary Google Fused Location Services client library (https://developers.google.com/location-context/fused-location-provider/) which is used to smartly combine location data from gps, wifi, cell bluetooth sources..
> 
> Only in Android 12 it became possible to use this service via a standard open-source Android Location API, see https://source.android.com/docs/automotive/location/coarse-location
> 
> So if OM switches to GPL then it won't be possible to ship this proprietary lib.

I think that OM has the right to ship their own binaries built from the GPL code it owns, and include this propietary lib. Other people don't. And obviously OM can't include GPL code from other projects.

**Reply by Atemu** [reply]
> I think that OM has the right to ship their own binaries built from the GPL code it owns, and include this propietary lib

Only if OM also has the rights to distribute the proprietary lib under the terms of the GPL which it cannot.

**Comment by pastk** [comment]
There is an argument that "permissive" licenses like Apache 2.0, MIT etc. are favored by commercial companies who might use the code without restrictions and without fear of legal issues and at some point are likely to voluntarily contribute their changes back to the open common codebase.

Practically speaking, in a decade-long history of MM/OM under Apache 2.0 there were/are closed source forks which grazed the commons and never contributed back upstream (please correct me if I'm wrong!) and then MM itself became closed source and stopped contributing back to the commons. And nowadays MM is the main competitor of the open source base (in the form of OM).

So it seems like OM being Apache 2.0 licensed didn't help the commons.

So OM might stay under Apache 2.0 license and hope for a miracle that some of closed forks will start contributing back (which never happened in ~10 years).
Or adopt a copyleft license and disadvantage closed forks.

**Reply by Atemu** [reply]
History has shown that commercial companies are most likely to act as leeches w.r.t. open source projects; profit off them and barely contribute back (if at all). If a commercial company wants to commit significant resources to an open source project, they've historically just taken the existing code and added proprietary changes on top; selling it as their product. (Obviously without contributing anything back too.)

Something like the GPL is the only choice if you don't want that to happen. GPL is like poison to leeches; there are leeches where it's strictly forbidden to have any GPL-licensed software on the development machine; no matter what it's used for. They're scared by it like vampires by garlic.

If inclusion in proprietary apps is a desired feature for some reason, the LGPL would be an option that, while keeping to door open quite a bit towards leeches, does restrict them in a few meaningful ways.  
I could see this for the map rendering component for instance; you may want to allow that to be included in proprietary apps as a library. The rest of the actual app would then be under the regular GPL though.

**Reply by pastk** [reply]
The map rendering component of Organic Maps is very tightly integrated with other components (its true for the most code in OM - little modularity and high integration and hence high optimization). So actually making it a library would require a substantial refactoring effort which might not worth it even..

It was touched here also: https://github.com/orgs/organicmaps/discussions/6003#discussioncomment-7017554

**Comment by pastk** [comment]
There might be contributors who would like to contribute under a copyleft license only. Hence using a permissive license keeps these contributors away.

A copyleft license allows incorporating code under permissive licenses. So effectively it welcomes all contributors.
